Make it optional to select parent categories

RelicCornhusk opened the pull request:
Based on the discussions on #54, I want to propose a change to how the listed transactions are filtered.

The idea is to make the second step of the flow (selecting parent categories where you have shared expenses) optional to also cater to users that don't track their shared expenses with categories. So if no parent categories are selected, a complete list of the transactions from the shared account gets listed (which to me seems more intuitive). Else, if one or more parent categories are selected, the current behaviour of the app is maintained (listing all transactions from the selected parent categories, regardless of selected account) to give users visibility of a transaction placed on the wrong account.

This PR invalidates #56 IMO, since its only purpose was to streamline the use for people who don't use categories to track shared expenses, so I'll be closing it.

chelseaSchmidt commented on client/appSrc/components/CostSharingForYnab.jsx:
```diff
@@ -282,11 +282,30 @@ const CostSharingForYnab = () => {
-        && transaction.approved
-        && !isTransactionATransfer(transaction)
-      ));
+      let displayedTransactions;
+
+      if (selectedParentCategories.length === 0) {
+        // No parent categories: filter by selected accounts only
+        displayedTransactions = transactionsSinceStartDate.filter((transaction) => (
+          isTransactionBeforeDate(transaction, endDate)
+          && transaction.approved
+          && !isTransactionATransfer(transaction)
+          && selectedAccounts.some((acct) => acct.accountId === transaction.account_id)
+        ));
+      } else {
+        // Parent categories selected: filter by category, ignore account
+        // Get all subcategory IDs from selected parent categories
+        const sharedCategoryIds = selectedParentCategories
+          .flatMap(({ subCategories }) => subCategories)
+          .map((cat) => cat.id);
+
+        displayedTransactions = transactionsSinceStartDate.filter((transaction) => (
+          isTransactionBeforeDate(transaction, endDate)
+          && transaction.approved
+          && !isTransactionATransfer(transaction)
+          && sharedCategoryIds.includes(transaction.category_id)
+        ));
+      }
```
I like where your head's at with this! There's a couple things to unpack here RE: code maintainability:

1. Let's follow the DRY principle. I think you mentioned you're relying solely on Copilot, so if you notice there's a lot of repeated code in what the AI spits out, try asking it via a follow-up prompt to update the code to be DRY. E.g. this is how I might rewrite this:
    ```
    const sharedCategoryIds = selectedParentCategories
      .flatMap(({ subCategories }) => subCategories)
      .map((cat) => cat.id);
    
    const displayedTransactions = transactionsSinceStartDate.filter((transaction) => (
      isTransactionBeforeDate(transaction, endDate)
      && transaction.approved
      && !isTransactionATransfer(transaction)
      && selectedParentCategories.length === 0
        ? selectedAccounts.some((acct) => acct.accountId === transaction.account_id)
        : sharedCategoryIds.includes(transaction.category_id)
    ));
    ```
2. This might make the above a moot point, depending on how you implement the change, but lets keep the new classifying logic isolated inside `classifyTransactions`. That will also prevent the need to repeat the following code block in two places, which would be another DRY win:
    ```
    const sharedCategoryIds = selectedParentCategories
        .flatMap(({ subCategories }) => subCategories)
        .map((cat) => cat.id);
    ```

chelseaSchmidt commented on client/appSrc/components/CostSharingForYnab.jsx:
```diff
@@ -492,7 +511,7 @@ const CostSharingForYnab = () => {
 
         <SectionContent>
-          <Subtitle>Select the YNAB parent category(ies) where you track shared expenses</Subtitle>
+          <Subtitle>(Optional) Select the YNAB parent category(ies) where you track shared expenses</Subtitle>
```
Let's update to `(Optional) If you track shared expenses under separate parent categories, select them here. This will turn on cross-checking for misclassified transactions.` It's wordy but makes the behavior change more obvious

chelseaSchmidt commented on client/appSrc/components/utils/classifyTransactions.js:
```diff
@@ -8,9 +8,20 @@ const classifyTransactions = ({
+  // If no parent categories selected, treat all as "in shared categories"
+  if (sharedCategoryIds.length === 0) {
+    return {
+      transactionsInSharedCategories: displayedTransactions,
```
The `transactionsInSharedCategories` property would need to be renamed with this implementation - the name should stay aligned with what data is stored in the array

RelicCornhusk replied:
Renamed it to filteredTransactions, let me know if this makes more sense. See [99f4c89](https://github.com/RelicCornhusk/cost-sharing-for-ynab/commit/99f4c89caa2fd64647128eda9d8a672295b6ae97)

chelseaSchmidt reviewed (commented):
Love this. Smart to make the behavior change implicit in how you make selections. Now the rest of the world can use it the intuitive way, while I continue using it my weird way 😂

RelicCornhusk commented:
hey @chelseaSchmidt! just rebased the branch and fixed some issues that came from the drift. I updated the commit references on my comments. I think I managed to migrate the filtering logic to classifyTransactions.js, but maybe it could still be DRYer? Not sure, let me know your thoughts.

RelicCornhusk commented:
I like it, it's cleaner! Thanks for the help 💪

chelseaSchmidt commented:
Sweet - this is deployed now 🎉
